### ISSUE TITLE
[Feature Request] Loops and user-defined events

### ISSUE THREAD

**Issue by AntoinePrv** [pytorch/ignite#217]
This one is more of a long shot. Today ignite provides two nested loops (epochs and batchs) with a custom callback callback function.

In some case however it requires more than two loops. For instance, in truncated backpropagation through time, one iterates over, epochs, then batches, then over time.

Another example I'm thinking about (without adding loops) is for someone working on optimization. Such a person may want to add an event between the `loss.backward()` and `optim.step()` and then swap different handles in their work without having to care bout the rest of the training loop.

One can have workarounds for both these examples, but I think it is valuable to have some way to let the user add some more loops and more events. I believe being very modular will let researchers more easily separate their code between what they are working on, and what is left untouched.

**Comment by vfdev-5** [comment]
@AntoinePrv thanks for the feature request ! Even if you are talking about some specific situations that can be implemented by a custom engine's `process_function`, maybe this can be implemented in common way with what you proposed in #215 - more engines.

In a `contrib` we could eventually provide some customized `process_function`s : 
- `process_function_time_iteration` that could have an internal loop over time
- `optimization_process_function` that could fire some custom events between `loss.backward()` and `optim.step()`
- ...

What do you think about ?

**Comment by AntoinePrv** [comment]
Yes, a `contrib` could be a nice solution to these two examples (and more to come hopefully!).
For the TBTT case, we can build the inner loop inside the `process_function` passed to the `Engine` and even fire custom events dedicated to this iteration.

I wonder how to fire custom events within the given framework.
More precisely:
  - [Events](https://github.com/pytorch/ignite/blob/c4563ddc67bee03fc616e87b84abd9b5e17e21cc/ignite/engine/engine.py#L13) is a `Enum`;
  - Engine [checks that an event is in the Enum](https://github.com/pytorch/ignite/blob/c4563ddc67bee03fc616e87b84abd9b5e17e21cc/ignite/engine/engine.py#L101) before adding it as an handler.

Intuitively, if `Events` was a `dict` (not implying this is what should be done), we could define new events in a `contrib` (or in user code), add it to the supported events, and then fire them and add handlers as usual. This would have no impact if the code is not used because such events would be added at runtime.

Any idea how to proceed?

**Comment by vfdev-5** [comment]
@AntoinePrv thinking about custom events, actually I do not see why we need to pass by event handling, we can call directly the handler in the processing function. For example,
```
def on_custom_event(engine, *args, **kwargs):
    pass

def update_with_custom_event(engine, batch):
    
    for t in range(time_range):
        # do smth
        # fire event -> just call the handler
        on_custom_event(engine)
        # do smth else
        
engine = Engine(update_with_custom_event)
```
What do you think ?

**Comment by AntoinePrv** [comment]
Well, why do we need events in the first place? They allow for modular and reusable code. Custom events would let contrib code present the same type of API as core ignite.

Ideally I was imagining a user interface like this:
```python
from ignite.engine import Events
from ignite.contrib.engine import create_supervised_tbptt_trainer
# Events are added automatically by importing this module

# ...

trainer = create_supervised_tbptt_trainer(model, optimizer, loss, tbptt_step)

@trainer.on(Events.TIME_ITERATION_COMPLETED)
def log_training_loss(trainer):
    # ...
```

**Comment by alykhantejani** [comment]
@AntoinePrv thanks for the feature request. I think you're correct - if we have a contrib, which will have more pre-canned engines, we should allow for these contrib engines to add to custom engines.

This should be relatively simple to do :). I can try to PR this over the weekend, unless you fancy sending a PR?
